fdb-extensions/src/main/java/com/apple/foundationdb/async/MoreAsyncUtil.java

```diff
@@ -857,6 +857,33 @@ public static void closeIterator(@Nonnull Iterator<?> iterator) {
         }
     }
 
+    /**
+     * Compose a handler bi-function to the result of a future. Unlike the
+     * {@link AsyncUtil#composeHandle(CompletableFuture, BiFunction)}, which completes exceptionally only when the
+     * <code>handler</code> completes exceptionally, it completes exceptionally even if the supplied action itself
+     * (<code>future</code>) encounters an exception.
+     * @param future future to compose the handler onto
+     * @param handler handler bi-function to compose onto the passed future
+     * @param exceptionMapper function for mapping the underlying exception to a {@link RuntimeException}
+     * @param <V> return type of original future
+     * @param <T> return type of final future
+     * @return future with same completion properties as the future returned by the handler
+     * @see AsyncUtil#composeHandle(CompletableFuture, BiFunction)
+     */
+    public static <V, T> CompletableFuture<T> composeWhenComplete(
+            @Nonnull CompletableFuture<V> future,
+            @Nonnull BiFunction<V,Throwable,? extends CompletableFuture<T>> handler,
+            @Nullable Function<Throwable,RuntimeException> exceptionMapper) {
```

Comment:
I *think* (and I could be wrong), but *think* that the semantics would be closer to `.whenComplete` if this returned a `CompletableFuture<V>` and the `handler` was a `BiFunction<V,Throwable,? extends CompletionStage<?>>` (or something like that--maybe a `CompletableFuture` instead of a `CompletionStage`, and maybe `Void` instead of `?`), then it would only use the returned future to create a call back, but it would return back the original result from the operation (if there was no error). It could also be the case that we don't actually care about replicating the semantics of `.whenComplete`

Reply:
Yes, but this method can provide more capabilities which may not be easy to implement. I've now wrapped this method to get  a `composeWhenComplete` which is like what you are suggesting. But I'd like to keep this method in a different name. It's now named `composeWhenCompleteAndHandle` but probably you have a better name for it.